### PR TITLE
refactor!: Update App Concurrency Model

### PR DESCRIPTION
**Pull request by jshlbrd**
## Description

- Refactors the core application's concurrency model to use [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup)
- Changes all application errors from public to private
- Changes the internal DynamoDB Query API to refer to partition keys as `:pk` and sort keys as `:sk`
  - This follows best practices for the DynamoDB key condition expression syntax
- Standardizes the system's naming conventions and design patterns

## Motivation and Context

For the concurrency model refactor, we've known for a while that the existing design has a risk of leaking goroutines when data processing errors happen, but fixing this wasn't a priority because we've never observed it as an issue since the system runs in AWS Lambda (over time the "serverless container" that the application runs in is reaped, cleaning up lingering goroutines). 

Now is as good a time as any to try and improve the design, so this PR replaces the Errs and Kill channels with an [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) and derived context. Functionally this is the same, but provides two advantages for reliability:
- we're no longer using buffered channels
- errors safely propagate across goroutines

The only significant change in this refactor is the introduction of a Channel struct that allows safer closing of and writing to channels. This was required to interrupt producer goroutines stuck on unbuffered channels when the consumer already terminated, but has an advantage of future proofing changes we might make to channel behavior.

References for these changes:
- https://go101.org/article/channel-closing.html
- https://www.fullstory.com/blog/why-errgroup-withcontext-in-golang-server-handlers/
- https://github.com/uber-go/guide/blob/master/style.md#channel-size-is-one-or-none

For the remaining changes, I think they are self-explanatory -- mostly focused on standardizing our approach to conventions across the system with supported documentation.

## How Has This Been Tested?

- Added a unit test for cmd/app that uses [Uber's goleak package](https://github.com/uber-go/goleak) to test for goroutine leaks
  - Unrelated to testing, cmd/app_test.go contains a fully functional Substation app that can be used as a toy example
- Tested on two large, complex deployments

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

* [x] My code follows the code style of this project.
* [x] My change requires a change to the documentation.
* [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by shellcromancer — APPROVED** [review]
Nice refactor!

File: process/process.go [hunk 1]
```diff
@@ -322,7 +319,7 @@ func newBatch(s *[]config.Capsule) []config.Capsule {
 	return make([]config.Capsule, 0, 10)
 }
 
-// conditionallyApplyBatch uses conditions to dynamically apply processors to a slice of encapsulated data. This is a convenience function for the ApplyBatch method used in most processors.
+// conditionallyApplyBatch uses conditions to dynamically processors to a slice of encapsulated data. This is a convenience function for the ApplyBatch method used in most processors.
```
Comment (julieagnessparks) [inline process/process.go:322]: nit: to dynamically apply processors

**Review by julieagnessparks — APPROVED** [review]
Looks great! Left a couple comments on typos in documentation but NBD

File: config/config_test.go [hunk 2]
```diff
@@ -114,9 +114,9 @@ func TestCapsuleDelete(t *testing.T) {
 		cap.SetData(test.data).SetMetadata(test.metadata)
 		cap.Delete(test.key)
 
-		if !bytes.Equal(cap.GetData(), test.dataExpected) &&
-			!bytes.Equal(cap.GetMetadata(), test.metadataExpected) {
-			t.Logf("expected %s %s, got %s %s", test.dataExpected, test.metadataExpected, cap.GetData(), cap.GetMetadata())
+		if !bytes.Equal(cap.Data(), test.dataExpected) &&
+			!bytes.Equal(cap.Metadata(), test.metadataExpected) {
+			t.Logf("expected %s %s, got %s %s", test.dataExpected, test.metadataExpected, cap.Data(), cap.Metadata())
```
[the remark this answers is not in the document]
Comment (jshlbrd) [reply config/config_test.go:119]: added as an issue: https://github.com/brexhq/substation/issues/31

**Pull request merged**